File: c/143-reorder-list/reorder-list.c
struct ListNode {
    int val;
    struct ListNode *next;
};

void merge(struct ListNode *list1, struct ListNode *list2){
    if(!list2)
        return;

    struct ListNode *temp1 = list1->next;
    struct ListNode *temp2 = list2->next;

    list1->next = list2;
    list2->next = temp1;

    merge(temp1, temp2);
}

struct ListNode *reverse(struct ListNode *head){
    struct ListNode *previous = NULL;
    struct ListNode *current = head;
    while(current != NULL){
        struct ListNode *next = current->next;
        current->next = previous;
        previous = current;
        current = next;
    }
    return previous;
}

void reorderList(struct ListNode* head) {

    if(!head || !head->next)
        return;

    struct ListNode *slow = head, *fast = head;
    while(fast && fast->next){
        slow = slow->next;
        fast = fast->next->next;
    }

    struct ListNode *current = slow->next;
    slow->next = NULL;

    struct ListNode *previous = reverse(current);

    merge(head, previous);
}

File: java/143-reorder-list/reorder-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {

    public void reorderList(ListNode head) {
        if (head == null || head.next == null) {
            return;
        }

        // 1. Find the middle node
        ListNode slowPointer = head;
        ListNode fastPointer = head;

        while (fastPointer.next != null && fastPointer.next.next != null) {
            slowPointer = slowPointer.next;
            fastPointer = fastPointer.next.next;
        }

        // 2. Reverse the right half of the linked list;
        ListNode prev = null;
        ListNode curr = slowPointer.next;

        slowPointer.next = null;

        while (curr != null) {
            ListNode next_node = curr.next;
            curr.next = prev;
            prev = curr;
            curr = next_node;
        }

        // 3. Merge two half of the linked list using two pointers from each
        ListNode first = head;
        ListNode second = prev;

        while (second != null) {
            ListNode nextNode1 = first.next;
            ListNode nextNode2 = second.next;

            first.next = second;
            second.next = nextNode1;

            first = nextNode1;
            second = nextNode2;
        }
    }
}
